kong/pdk/cluster.lua

```diff
@@ -0,0 +1,85 @@
+--- Node-level utilities
+--
+-- @module kong.node
+
+local utils = require "kong.tools.utils"
+
+
+local kong = kong
+local CLUSTER_ID_CACHE_KEY = require("kong.constants").CLUSTER_ID_CACHE_KEY
+
+
+local function fetch_cluster_id()
+  local res, err = kong.db.parameters:select({ key = CLUSTER_ID_CACHE_KEY, })
+  if err then
+    return nil, err
+  end
+
+  if res then
+    return res.value
+  end
+
+  local cluster_id = utils.uuid()
+
+  res, err = kong.db:cluster_mutex(CLUSTER_ID_CACHE_KEY, nil, function()
+    -- another worker might have already generated the cluster ID, double check
+    -- in here before overwriting it
+
+    res, err = kong.db.parameters:select({ key = CLUSTER_ID_CACHE_KEY, })
+    if err then
+      error(err)
+    end
+
+    if not res then
+      assert(kong.db.parameters:insert({ key = CLUSTER_ID_CACHE_KEY,
+                                         value = cluster_id, }))
+
```

Comment:
Question for personal edification: why is the cluster ID generation done in lazily and not as part of migrations?

Reply:
Good point. The reason why it was generated lazily was because the first version of this PR only worked under Hybrid mode and we did not have needs to generate them for traditional deployment (wouldn't hurt either though). We can generate it in the migrations and it will reduce the logic in the PDK function by a lot. I will make these changes.